tools/scoring/score/dimensions/entity_point_identification.py

```diff
@@ -0,0 +1,159 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+""" Core component """
+
+from score.dimensions.dimension import Dimension
+from score.types_ import DeserializedFile, EntityInstance, CloudDeviceId, PointsVirtualList
+from score.constants import FileTypes
+
+from typing import Set
+
+PROPOSED, SOLUTION = FileTypes
+
+
+class EntityPointIdentification(Dimension):
+  """
+  Quantifies whether the proposed file
+  included the correct points in each entity.
+  """
+  def _isolate_entities_virtual(self,
+                                file: DeserializedFile) -> Set[EntityInstance]:
+    return set(
+        filter(self.is_entity_canonical,
+               filter(self.is_entity_virtual, file.values())))
+
+  def _fetch_points_virtual(
+      self, file: DeserializedFile,
+      entities_virtual: Set[EntityInstance]) -> PointsVirtualList:
+    # For each canonically typed virtual entity
+    # create a tuple containing a set and the entity's type.
+    # (The type is required by the matching algo for a different dimension.)
+    # For each link in the entity,
+    # if the field exists at the source
+    # add its raw field name to the set.
+    # Filter out sets which have no items
+    # and sort by number of fields represented in descending order.
+    # TODO: move the filtering/sorting functionality to add clarity.
+    return sorted(
+        list(
+            filter(
+                lambda tup: len(tup[0]) > 0,  # (raw field names, entity type)
+                [(set(file[link.source].translation[target_field].raw_field_name
+                      for target_field, source_field in link.field_map.items()
+                      for link in entity.links
+                      if target_field in file[link.source].translation),
+                  entity.type) for entity in entities_virtual
+                 for link in entity.links])),
+        key=lambda tup: len(tup[0]),  # (raw field names, entity type)
+        reverse=True)
+
+  def _isolate_entities_reporting(
+      self, file: DeserializedFile) -> Set[EntityInstance]:
+    return set(
+        filter(self.is_entity_canonical,
+               filter(self.is_entity_reporting, file.values())))
+
+  def _fetch_source_ids_virtual(
+      self, file: DeserializedFile,
+      entities_virtual: Set[EntityInstance]) -> Set[CloudDeviceId]:
+    # Aggregate IDs for entities which comprise the composites
+    # evaluated above. These will be filtered out so as to
+    # not be scored again.
+    return set(file[source].cloud_device_id for sublist in (
+        (link.source for target_field, source_field in link.field_map.items()
+         for link in entity.links
+         if target_field in file[link.source].translation)
+        for entity in entities_virtual for link in entity.links)
+               for source in sublist)
+
+  def evaluate(self):
```

Comment:
can you please add docstring here.

Reply:
85c589b8964d292819a342cd0c3dd7006e3705d9

Reply:
85c589b8964d292819a342cd0c3dd7006e3705d9